### PR TITLE
Fix Firefox console warnings.

### LINKED ISSUE #5912: Firefox WebGL warnings in console
When I load [the basic viewer](https://cesiumjs.org/Cesium/Build/Apps/CesiumViewer/index.html?stats=true) in Firefox (56.0, 64-bit, Ubuntu 16.04), I see repeated errors in the console:

> Error: WebGL warning: texImage2D: Alpha-premult and y-flip are deprecated for non-DOM-Element uploads.

When I try to work with [billboards](https://cesiumjs.org/Cesium/Apps/Sandcastle/index.html?src=Billboards.html&label=Showcases), I also see

> Error: WebGL warning: texSubImage2D: Texture has not been initialized prior to a partial upload, forcing the browser to clear it. This may be slow.

and

> Error: WebGL warning: texSubImage2D: This operation requires zeroing texture data. This is slow.

In my (much more complex) application, I get tons of these warnings, until Firefox just gives up and stops warning.  I don't know if this is actually having a performance impact; Chromium does do a few FPS better than Firefox but that could be for other reasons.

(I have seen #1317 but if you click through to the Bugzilla link, they think that issue has been fixed, and would only account for one of the 3 warning types.)

### PR DESCRIPTION
**Pull request by bagnell**
* Only use pre-multiply alpha and y flip for DOM sources. 
* Initialized textures before using texSubImage2D.
    * Use texImage2D when the sub image is the same size as the texture
    * Use texImage2D to initialize to zero when the sub image is smaller.

Fixes #5912

### CODE REVIEW

**Comment by cesium-concierge** [comment]
@bagnell, thanks for the pull request! Maintainers, we have a signed CLA from @bagnell, so you can review this at any time.

I noticed that [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md) has not been updated. If this change updates the public API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and comment on this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).

---

__I am a bot who helps you make Cesium awesome!__ Thanks again.

File: Source/Renderer/CubeMap.js [hunk 1]
```diff
@@ -121,25 +120,30 @@ define([
         gl.activeTexture(gl.TEXTURE0);
         gl.bindTexture(textureTarget, texture);
 
-        function createFace(target, sourceFace) {
+        function createFace(target, sourceFace, preMultiplyAlpha, flipY) {
+            // TODO: gl.pixelStorei(gl._UNPACK_ALIGNMENT, 4);
```
Comment (mramato) [inline Source/Renderer/CubeMap.js]: Did you mean to leave this TODO in?

Comment (bagnell) [reply Source/Renderer/CubeMap.js]: Yes, that was in the original code. I just moved it.

File: Source/Renderer/CubeMap.js [hunk 2]
```diff
@@ -121,25 +120,30 @@ define([
...
+            gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, false);
```
Comment (pjcozzi) [inline Source/Renderer/CubeMap.js]: Are there duplicate `pixelStorei` calls if `sourceFace.arrayBufferView` below is not defined?  If so, can we move these to the first `if` block?

File: Source/Renderer/CubeMapFace.js [hunk 3]
```diff
@@ -91,15 +96,58 @@ define([
...
+        gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, false);
+
         gl.activeTexture(gl.TEXTURE0);
         gl.bindTexture(target, this._texture);
 
-        if (source.arrayBufferView) {
-            gl.texSubImage2D(this._targetFace, 0, xOffset, yOffset, source.width, source.height, this._pixelFormat, this._pixelDatatype, source.arrayBufferView);
-        } else {
-            gl.texSubImage2D(this._targetFace, 0, xOffset, yOffset, this._pixelFormat, this._pixelDatatype, source);
+        var uploaded = false;
```
Comment (pjcozzi) [inline Source/Renderer/CubeMapFace.js]: Here and below seems more complicated than it needs to be; I will not have the time to dive deep for awhile.

@lilleyse can you take a look?

**Comment by mramato** [comment]
@bagnell should we get this into tomorrow's release?  Seems like we would want to (especially if it's a speed improvement).

**Comment by pjcozzi** [comment]
Please do not rush the review - would rather see this go into the following release.

**Comment by mramato** [comment]
Do you think this will help with #3888?  The Firefox tab now crashes if you run the tests with WebGL validation.

**Comment by bagnell** [comment]
No, the tab still crashes for me.

**Comment by thw0rted** [comment]
Hey all, just curious if this fell off somebody's plate -- I'm still seeing the warnings I mentioned in #5912 .

**Comment by hpinkos** [comment]
Thanks for the reminder @thw0rted 

@bagnell what's the status of this PR?

**Comment by bagnell** [comment]
Waiting for a review from @lilleyse.

File: Source/Renderer/CubeMap.js [hunk 4]
```diff
@@ -121,25 +120,30 @@ define([
...
-            createFace(gl.TEXTURE_CUBE_MAP_POSITIVE_Z, source.positiveZ);
-            createFace(gl.TEXTURE_CUBE_MAP_NEGATIVE_Z, source.negativeZ);
+            createFace(gl.TEXTURE_CUBE_MAP_POSITIVE_X, source.positiveX, preMultiplyAlpha, flipY);
+            createFace(gl.TEXTURE_CUBE_MAP_NEGATIVE_X, source.negativeX, preMultiplyAlpha, flipY);
+            createFace(gl.TEXTURE_CUBE_MAP_POSITIVE_Y, source.positiveY, preMultiplyAlpha, flipY);
+            createFace(gl.TEXTURE_CUBE_MAP_NEGATIVE_Y, source.negativeY, preMultiplyAlpha, flipY);
+            createFace(gl.TEXTURE_CUBE_MAP_POSITIVE_Z, source.positiveZ, preMultiplyAlpha, flipY);
+            createFace(gl.TEXTURE_CUBE_MAP_NEGATIVE_Z, source.negativeZ, preMultiplyAlpha, flipY);
         } else {
             gl.texImage2D(gl.TEXTURE_CUBE_MAP_POSITIVE_X, 0, pixelFormat, size, size, 0, pixelFormat, pixelDatatype, null);
```
[the remark this answers is not in the document]
Comment (bagnell) [reply Source/Renderer/CubeMap.js]: I did it this way for the case where you create a texture and upload the entire texture later. This avoids creating an array buffer of all zeroes. The only time the array buffer of zeroes would be created would be when a texture is created and then a bunch of `texSubImage2D`'s are called to fill out an atlas. If we don't care about that case or optimization, then Ill change this.

File: Source/Renderer/Texture.js [hunk 5]
```diff
@@ -195,11 +197,16 @@ define([
...
+                // Only valid for DOM-Element uploads
+                gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, preMultiplyAlpha);
+                gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, flipY);
+
                 // Source: ImageData, HTMLImageElement, HTMLCanvasElement, or HTMLVideoElement
                 gl.texImage2D(textureTarget, 0, internalFormat, pixelFormat, pixelDatatype, source);
             }
         } else {
             gl.texImage2D(textureTarget, 0, internalFormat, width, height, 0, pixelFormat, pixelDatatype, null);
+            initialized = false;
```
[the remark this answers is not in the document]
Comment (bagnell) [reply Source/Renderer/Texture.js]: Same comment here.

File: Source/Renderer/CubeMap.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (bagnell) [reply Source/Renderer/CubeMap.js]: Also, a texture that is going to be rendered to cannot be set.

**Comment by bagnell** [comment]
@lilleyse This is ready for another look.

File: Source/Renderer/Texture.js [hunk 6]
```diff
@@ -469,15 +477,44 @@ define([
         var target = this._textureTarget;
 
         // TODO: gl.pixelStorei(gl._UNPACK_ALIGNMENT, 4);
-        gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, this._preMultiplyAlpha);
-        gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, this._flipY);
+        gl.pixelStorei(gl.UNPACK_PREMULTIPLY_ALPHA_WEBGL, false);
+        gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, false);
```
Comment (lilleyse) [inline Source/Renderer/Texture.js]: Make the same changes as `CubeMapFace` so that `pixelStorei` isn't called multiple times in some cases

**Comment by lilleyse** [comment]
Tests pass and the Sandcastle demos I tried still work.

> Here and below seems more complicated than it needs to be; I will not have the time to dive deep for awhile. @lilleyse can you take a look?

It seems mostly unavoidable to me based on this discussion https://github.com/AnalyticalGraphicsInc/cesium/pull/5939#discussion_r169324330.

Do we want to release this with 1.43? @pjcozzi did you want to take a look?

**Comment by pjcozzi** [comment]
I don't have bandwidth to look at this for 1.43, but please merge if you are confident.

File: Source/Renderer/Texture.js (hunk 6 above)
Comment (bagnell) [reply Source/Renderer/Texture.js]: Done.

**Comment by markerikson** [comment]
I'm still seeing a related warning with Cesium 1.45 and Firefox 52 (which we are unfortunately stuck on for our application):

> Error: WebGL: texSubImage2D: Conversion requires pixel reformatting.

Was that supposed to be resolved by this PR?

**Comment by hpinkos** [comment]
Hi @markerikson, do you see that error in Cesium 1.44 as well?  This PR was added in 1.43 so if you're seeing the error in 1.45 but not 1.44 it means that we might have introduced something in 1.45

**Comment by markerikson** [comment]
Loading up our app with 1.44, I actually see considerably more warnings:

> Error: WebGL: texImage2D: Alpha-premult and y-flip are deprecated for non-DOM-Element uploads.
> Error: WebGL: texImage2D: Conversion requires y-flip.
> Error: WebGL: texSubImage2D: Texture has not been initialized prior to a partial upload, forcing the browser to clear it. This may be slow.
> Error: WebGL: texSubImage2D: This operation requires zeroing texture data. This is slow.
> Error: WebGL: texSubImage2D: Conversion requires pixel reformatting.

(All repeated quite a few times, just isolating the unique messages.)

In 1.45, I only see the "pixel formatting" message.

**Comment by bagnell** [comment]
I haven't seen that error, but I'm using Firefox 60. Do you see that in the Viewer or any Sandcastle example? If not, can you post an example so I can take a look?

**Comment by markerikson** [comment]
Yes, just loading up the main Sandcastle page in FF52 immediately shows the "pixel reformatting" warning in the console.

**Comment by hpinkos** [comment]
Hmm, I don't see it on FF59 either.  It's unfortunate you're unable to upgrade

**Comment by markerikson** [comment]
Yeah, we have an internal app that targets a specific Firefox ESR version at a time, so we're still targeting FF52 for this dev cycle.  We'll live with it, just was wanting to see if it was supposed to have been resolved already (and at least there's fewer warnings).

**Pull request merged**